### PR TITLE
Enhance API-Testing-Guidelines.md

### LINKED ISSUE #158: Enhancement of the Testing Guidelines
**Problem description**
The first version of the  has to be enhanced with more detailed instructions to have consistent Test Plans across the WGs.

**Possible evolution**

* Enhance the [API testing guidelines](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-Testing-Guidelines.md) with more detailed instructions.

* Build necessary artifacts to support testing plans 

**Additional context**

We include here a draft of the proposal, to trigger the discussion. When we reach enough consensus on the approach we can create a PR with the modifications.

----

# Proposal

Testing implementations can use the Gherkin feature files with two different approaches:

- For testing automation with some framework that take Gherkin as an input, such as Cucumber or Behave. For those tools, the steps have to be written in a way that allows to map them to a function with some optional arguments.

- To be interpreted by a human as input for codification with another tool, which does not link to Gherkin natively, such as Postman, SoapUI, etc. In these cases, it is important that the scenario as a whole is unambiguous, and states how to build the request and validate the response.

### Feature structure

A feature file will typically test an API operation and will consist of several scenarios testing the behaviour of the API operation with different conditions or input content, validating that the response complies with the expected HTTP status, and that the response body meeets the expected JSON schema and some properties have the expected values.

### Configuration variables

Most scenarios will test a request and its response. Commonly, values to fill request bodies will not be known in advance as they will be specific for the test environment, and will have to be provided as a separate set of configuration variables.

A first stage is to identify those variables, e.g. device identifiers (phone numbers, IP addresses), status of a testing device, etc. How those variables are set and feed into the testing execution will depent on the testing tool (Postman environment, context in Behave, etc)

In order to pass information between steps and make use of the variables, we have to agree on some syntaxis to refer to those variables. For example, Postman uses `{{variable}}`, Gherkin uses `<variable>` for Scenario Outline, but this is not properly an Scenario Outline case. As a proposal, we may use something like [CONFIG:var]. 

Example:

```
  Scenario: Description
    Given the configuration variables:
      | variable  | description                           |
      | device    | Object identifying a device           |
    And the request body:
      """
      {
        "device": [CONFIG:device],
        ...
      }
    When the HTTP "POST" request is sent
    Then the response status code is "200"
    And the response JSON field "$.device" is "[CONFIG:device]"
```

A background section at the beginning of the feature file may set common configuration variables and attributes for all scenarios, e.g. `apiServer`, `baseUrl`, `resource`, etc

### Request setup

Typically the scenario will have to setup the request as part of the Given steps, filling the necessary path, query, header or body parameters. The guidelines can define a set of reusable steps for this, e.g.

```
- the resource "{url_path}"
- the resource “{url_path_str_format}” is parameterized with values:
    | param_name | param_value |
    | ---------- | ----------- |
    | xxx        | yyy         |
- the header "{name}" is set as "{value}"
- the headers:
    | param_name | param_value |
    | ---------- | ----------- |
    | xxx        | yyy         |
- the query parameter "{name}" is set as "{value}"
- the query parameters:
    | param_name | param_value |
    | ---------- | ----------- |
    | xxx        | yyy         |
- the request body field "{json_path" is set as "{value}"
- the request body:
    ```
    {
      "xxx": "yyy"
    }
    ```
```

### Request sending

Usually one When step will only be necessary:

```
When the HTTP "{method}" request is sent
```

For complex scenarios concatenating several requests, subsequent requests will usually be included in Then steps after the response to the first one is validated.

### Response validation

Several Then steps can validate the response. Some may be quite common, e.g.:

```
- the response status code is "200"
- the respone complies with the JSON schema at "{OAS_schema_path}"
- the response JSON field "{json_path}" is "{value}"
```

Others may be quite specific for the API logic and have to be designed ad-hoc.

## Open points

Some to start with...

### Device identification

Identification of the device. There are many possible combinations that comply with the schema `Device`, and there are no clear guidelines about which ones are required in order to pass the certification. 

- Do we have to test many combinations and validate which ones are passed in certain environment?

- Do we have to assume that `device` as an object is a configuration variable and each environment will provide a valid one? 

This topic links with the discussion in https://github.com/camaraproject/Commonalities/issues/127 

### Scope/reach of the testing plan

Does then Test plan have to test only that the interface complies with the spec, or does it have to test that the service is correctly provided? For example:

- Test that a device location is correctly verified or just that a valid response gets a 200, even if `verificationResult` value is not correct?

- Test that a device status for roaming is correct?

### PR DESCRIPTION
**Pull request by jlurien**
#### What type of PR is this?

* documentation

#### What this PR does / why we need it:

Enhances API testing guidelines with much more details and explanations. Initial version was too basic to assure consistent and coherent testing plans.

#### Which issue(s) this PR fixes:

Fixes #158 

#### Special notes for reviewers:

This is a first draft following proposal and initial feedback in  #158. 

#### Changelog input

```
* Enhancement of API-Testing-Guidelines

```

### CODE REVIEW

File: documentation/API-Testing-Guidelines.md [hunk 1]
```diff
@@ -1,40 +1,284 @@
...
+* the query parameter "startDate" is in the past
+* the request body property "$.device" is a valid device which location is known
+```
+
+Testing plans should identify the list of input values that have to be provided by the implementation.
+
+
+### Request sending <a name="request-sending"></a>
+
+For scenarios testing individual operations, the request will be send by means of a single `When` step, indicating the `operationId` specified in the spec, as the `operationId` uniquely identifies the HTTP method and resource path. E.g.:
```
Comment (PedroDiez) [inline documentation/API-Testing-Guidelines.md:161]:
```suggestion
For scenarios testing individual operations, the request will be sent by means of a single `When` step, indicating the `operationId` specified in the spec, as the `operationId` uniquely identifies the HTTP method and resource path. E.g.:
```

File: documentation/API-Testing-Guidelines.md [hunk 2]
```diff
@@ -1,40 +1,284 @@
...
+  - [CRUD operations](#crud-operations)
+  - [Notifications](#notifications)
+  - [Error scenarios](#error-scenarios)
+- [References](#references)
 
 ## Introduction <a name="introduction"></a>
 This document captures guidelines for the API testing in CAMARA project. These guidelines are applicable to every API to be worked out under the CAMARA initiative.
 
 ## Test Cases contribution for APIs <a name="contribution"></a>
 Based on the decision taken in the Commonalities working group and as documented in [API-Readiness-Checklist](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-Readiness-Checklist.md), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases and documentation. If subprojects also intend to add test implementations, an aligned single implementation that is agreed among all provider implementors could be added to the main subproject repo. If no alignment is possible, each provider implementer will add the test implementation to their own repos.
```
Comment (rartych) [inline documentation/API-Testing-Guidelines.md:26]: The updated API Readiness Checklist is proposed in Release Msanagement: https://github.com/camaraproject/ReleaseManagement/pull/29

It impacts this section and possibly "Basic API test cases & documentation" and "Enhanced API test cases & documentation"  should be defined here.

File: documentation/API-Testing-Guidelines.md [hunk 3]
```diff
@@ -1,40 +1,284 @@
...
+
+`https://github.com/camaraproject/DeviceLocation/blob/main/code/Test_definitions/location-verification.feature`
+
 
-## Best practices and recommendations <a name="recommendations"></a>
+Filename:
+
+- APIs with a single feature file will use the api-name as file name: E.g. `location-verification.feature`
+- For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the `operationId` as file name: `createSession.feature`.
+- For APIs with several feature files, not split by operation, may use the api-name along some other description, e.g.: `qod-notifications.feature`
```
[the remark this answers is not in the document]
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:43]: Split is right I think :)

File: documentation/API-Testing-Guidelines.md (hunk 2 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:26]: It's difficult to set the difference between Basic and Enhanced. To me basic tests cases should cover at least that the implementation interface is compliant with the API spec, in terms of syntax and support for all mandatory aspects. Enhanced test plan may validate unusual error scenarios and corner cases. But this is very subtle.

**Comment by jlurien** [comment]
In the latter commit:

Proposal updated with comments and feedback after dedicated session on May 16th.

- Tests coverage section moved to to the top, merging legacy section "Test Cases contribution for APIs".
- Added distinction between basic and enhanced test plan, to be aligned with Release Management proposal. There is no proper definition of what is basic or enhanced, so I suggested some initial test.
- Added explicitly that environment variables may be kept in a separated file. We should enhance this part with more details and a suggested template.

Please review these changes.

File: documentation/API-Testing-Guidelines.md [hunk 4]
```diff
@@ -1,41 +1,288 @@
...
+## Test plan coverage <a name="coverage"></a>
 
-## Location of feature file <a name="location"></a>
-The feature file will reside under 
-```
-Subproject_Repository/code/Test_definitions/file_name.feature
-for e.g. https://github.com/camaraproject/QualityOnDemand/blob/main/code/Test_definitions/QoD_API_Test.feature
-```
+Based on the decision taken in the Release Management and Commonalities working groups, and as documented in [API-Readiness-Checklist](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-Readiness-Checklist.md), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases and documentation:
```
Comment (rartych) [inline documentation/API-Testing-Guidelines.md:26]: I propose to change the link to Release Management wiki.
Current API Readiness Checklist will be deprecated: https://github.com/camaraproject/Commonalities/issues/202
```suggestion
Based on the decision taken in the Release Management and Commonalities working groups, and as documented in [API Release Process](https://wiki.camaraproject.org/display/CAM/API+Release+Process)](https://wiki.camaraproject.org/display/CAM/API+Release+Process#APIReleaseProcess-APIreadinesschecklist), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases and documentation:
```

File: documentation/API-Testing-Guidelines.md [hunk 5]
```diff
@@ -1,41 +1,288 @@
...
+If subprojects also intend to add test implementations, an aligned single implementation that is agreed among all provider implementors could be added to the main subproject repo. If no alignment is possible, each provider implementer will add the test implementation to their own repos.
+
+* Test plans are not expected to certify or measure the quality of the implementation in terms of accuracy or veracity of the response, but to validate that the interface is compliant with the API specification and that the API is fully implemented.
+
+* When the API allows options or alternative approaches, the test plan should cover those situations and define the expected behaviour for inputs or options not supported.
+  - Implementations will need to inform in advance to the API tester about which subset of options are supported. 
+
+### Independent operations <a name="independent-operations"></a>
+
+Operations that execute an action or retrieve information from the operator, without modifying a resource, will be tested in its own dedicated feature file.
```
Comment (mdomale) [inline documentation/API-Testing-Guidelines.md:40]: Does this mean we will need two different feature files for QoD Operations ? If yes then what would be significance of it ?

File: documentation/API-Testing-Guidelines.md [hunk 6]
```diff
@@ -1,41 +1,288 @@
...
+### Independent operations <a name="independent-operations"></a>
+
+Operations that execute an action or retrieve information from the operator, without modifying a resource, will be tested in its own dedicated feature file. 
+
+Test plan must validate that providing a valid input, a success response is received and that the response body complies with the JSON-schema in the API spec.
+
+For operations receiving an input request body with optional properties, several scenarios should be included, testing different sets of valid inputs.
+
+
+### CRUD operations <a name="crud-operations"></a>
```
Comment (mdomale) [inline documentation/API-Testing-Guidelines.md:47]: Can we rename this to Resource Modifying Operations to make the differentiation clearer

File: documentation/API-Testing-Guidelines.md [hunk 7]
```diff
@@ -1,41 +1,288 @@
...
+
+For path resources with CRUD (Create/Read/Update/Delete) operations, some scenario should validate that the new state of the resource is coherent with the response to the operation.
```
Comment (mdomale) [inline documentation/API-Testing-Guidelines.md:49]: * response -> request correction

File: documentation/API-Testing-Guidelines.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:26]: We can clarify that explicitly that, but the "API Readiness Checklist" applies to each API individually, so yes, each API which is part of a meta-release, regardless whether it is sharing a subproject with other APIs, must have its own test plan.

File: documentation/API-Testing-Guidelines.md (hunk 5 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:40]: QoD is mostly a CRUD API, encapsulating sessions resources, so it lies on the other category. Anyway, it is up to QoD subproject, and the contributors of its test plans, to decide on how many feature files are convenient. Splitting a test plan in several features is a test plan design decision, taking into account the complexity of the test plan. The more complex a test plan is, the most likely is that it will be split into several feature files. As in any software project.

These guidelines are recommendations for the design, so we may make the text less normative: "will be tested in its own dedicated feature file" -> "will be **usually** tested in its own dedicated feature file"

File: documentation/API-Testing-Guidelines.md (hunk 6 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:47]: Fine to me

File: documentation/API-Testing-Guidelines.md (hunk 7 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:49]: I can clarify this further, but it really meant the **response**. The request may ask the server for something, but it is the response which tells the client whether the server has granted the request and therefore modify the state of the resource accordingly.

File: documentation/API-Testing-Guidelines.md (hunk 5 above)
Comment (mdomale) [reply documentation/API-Testing-Guidelines.md:40]: @jlurien  OK with us kindly accommodate this in PR file as well

File: documentation/API-Testing-Guidelines.md (hunk 6 above)
Comment (mdomale) [reply documentation/API-Testing-Guidelines.md:47]: @jlurien  Kindly accommodate in PR file as well

File: documentation/API-Testing-Guidelines.md (hunk 7 above)
Comment (mdomale) [reply documentation/API-Testing-Guidelines.md:49]: @jlurien It seems new state check is missing in the detailed steps below

File: documentation/API-Testing-Guidelines.md [hunk 8]
```diff
@@ -1,41 +1,288 @@
...
+
+* Check that when a webhook is provided, the expected events are received in the `notificationUrl`, with the right `notificationAuthToken`, for those situations specified in the API.
+* If the API allows to update a previously provided webhook:
+  - If the `notificationUrl` is modified, validate that events are received in the modified value.
+  - If the `notificationUrl` can be nullified, validate that events are not longer received.
+  - If the `notificationAuthToken` is modified, validate that the new token is used for the events.
+
+For the explicit subscriptions model:
+
+* Validate that the subscribed events are received in the `notificationUrl`, with the right `notificationAuthToken`, for those situations specified in the API.
```
Comment (mdomale) [inline documentation/API-Testing-Guidelines.md:74]: @jlurien  Consideration of new template  event-subscription-template.yaml needed in Testing-Guidelines as well

File: documentation/API-Testing-Guidelines.md [hunk 9]
```diff
@@ -1,41 +1,288 @@
...
+* For subscriptions that provide `subscriptionExpireTime`, validate that the subscribed events are not longer received after the expiration time.
+* For subscriptions that provide `subscriptionMaxEvents`, validate that the subscribed events are not longer received after the maximum events limit is reached..
+* Validate that after a subscription is deleted, the subscribed events are not longer received.
+
+### Error scenarios <a name="error-scenarios"></a>
+
+* All errors explicitly documented in the API spec must be covered by one or more dedicated error scenarios.
+  - In particular operations that expect a `device` object allowing one or more identifiers to be provided, must include scenarios to test all error cases defined.
```
Comment (mdomale) [inline documentation/API-Testing-Guidelines.md:82]: @jlurien Can you please clarify why only device object is considered explicitly? We think generic phrase would be fine for covering error scenarios.

File: documentation/API-Testing-Guidelines.md [hunk 10]
```diff
@@ -1,41 +1,288 @@
...
+E.g.: 
+
+`https://github.com/camaraproject/DeviceLocation/blob/main/code/Test_definitions/location-verification.feature`
+
+
+Filename:
+
+- APIs with a single feature file will use the api-name as file name: E.g. `location-verification.feature`
+- For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the `operationId` as file name: `createSession.feature`.
+- For APIs with several feature files, not split by operation, may use the api-name along some other description, e.g.: `qod-notifications.feature`
```
Comment (mdomale) [inline documentation/API-Testing-Guidelines.md:116]: @jlurien  We think API name should be mandatory with some added description if needed .

File: documentation/API-Testing-Guidelines.md (hunk 8 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:74]: I don't get this comment, sorry. Could you please elaborate further? Which template and which considerations?

File: documentation/API-Testing-Guidelines.md (hunk 9 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:82]: Because recently errors for device processing have been discussed specifically, and new design guidelines also have a dedicated mention to device error management. But this do not exclude any other error scenario. Idea is to give an example of deep of tests that are expected,

File: documentation/API-Testing-Guidelines.md (hunk 10 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:116]: Fine to me, I could suggest having always the api-name and, when the testing plan is composed of several files, that suffix, which can be operationId or other description.

File: documentation/API-Testing-Guidelines.md (hunk 8 above)
Comment (mdomale) [reply documentation/API-Testing-Guidelines.md:74]: @jlurien https://github.com/camaraproject/Commonalities/pull/189 This was new template introduced & consideration of this for Testing guidelines will be useful cc @akoshunyadi

Comment (rartych) [reply documentation/API-Testing-Guidelines.md:74]: @jlurien The template for explicit event subscription (like geofencing): https://github.com/camaraproject/Commonalities/blob/main/artifacts/camara-cloudevents/event-subscription-template.yaml

File: documentation/API-Testing-Guidelines.md [hunk 11]
```diff
@@ -1,41 +1,288 @@
...
+If subprojects also intend to add test implementations, an aligned single implementation that is agreed among all provider implementors could be added to the main subproject repo. If no alignment is possible, each provider implementer will add the test implementation to their own repos.
+
+* Test plans are not expected to certify or measure the quality of the implementation in terms of accuracy or veracity of the response, but to validate that the interface is compliant with the API specification and that the API is fully implemented.
+
+* When the API allows options or alternative approaches, the test plan should cover those situations and define the expected behaviour for inputs or options not supported.
+  - Implementations will need to inform in advance to the API tester about which subset of options are supported. 
+
+### Independent operations <a name="independent-operations"></a>
+
+Operations that execute an action or retrieve information from the operator, without modifying a resource, will be tested in its own dedicated feature file.
```
Comment (rartych) [inline documentation/API-Testing-Guidelines.md:40]:
```suggestion
Operations that execute an action or retrieve information from the operator, without modifying a resource, will be usually tested in its own dedicated feature file. 
```

File: documentation/API-Testing-Guidelines.md [hunk 12]
```diff
@@ -1,41 +1,288 @@
...
+### Independent operations <a name="independent-operations"></a>
+
+Operations that execute an action or retrieve information from the operator, without modifying a resource, will be tested in its own dedicated feature file. 
+
+Test plan must validate that providing a valid input, a success response is received and that the response body complies with the JSON-schema in the API spec.
+
+For operations receiving an input request body with optional properties, several scenarios should be included, testing different sets of valid inputs.
+
+
+### CRUD operations <a name="crud-operations"></a>
```
Comment (rartych) [inline documentation/API-Testing-Guidelines.md:47]:
```suggestion
### Resource modifying operations <a name="crud-operations"></a>
```

File: documentation/API-Testing-Guidelines.md (hunk 10 above)
Comment (rartych) [inline documentation/API-Testing-Guidelines.md:116]:
```suggestion
- For APIs with several feature files, not split by operation, should use the api-name along some other description, e.g.: `qod-notifications.feature`
```

File: documentation/API-Testing-Guidelines.md (hunk 8 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md]: It should be aligned now after @PedroDiez's suggestions. Please double check

**Review by PedroDiez — DISMISSED** [review]
Some minor editorials to be reviewed
Fine at content level

LGTM in advance

File: documentation/API-Testing-Guidelines.md (hunk 10 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:116]: I rephrased the sentence

File: documentation/API-Testing-Guidelines.md (hunk 5 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:40]: Added

File: documentation/API-Testing-Guidelines.md (hunk 7 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:49]: Text rephrased

File: documentation/API-Testing-Guidelines.md (hunk 9 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:82]: added reference to the device error management section of the API design guidelines

File: documentation/API-Testing-Guidelines.md (hunk 10 above)
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:116]: Suggestion from @rartych applied

**Comment by jlurien** [comment]
I think I have incorporated all suggestions and feedback, please double check

**Review by PedroDiez — APPROVED** [review]
LGTM

File: documentation/API-Testing-Guidelines.md (hunk 10 above)
Comment (mdomale) [reply documentation/API-Testing-Guidelines.md]: @jlurien  File still does not contain apiname-operationid/description in 2nd point  for Createsession.feature.
Kindly amend in the PR

**Comment by mdomale** [comment]
@jlurien  Except above 1 comment PR looks fine to us .

File: documentation/API-Testing-Guidelines.md (hunk 10 above)
Comment (rartych) [reply documentation/API-Testing-Guidelines.md]: I think adding api-name in the second point brings consistency (filename always starts with api-name) , but can be helpful in the future with multiple API specifications in the same repository (distinguishing tests in case the same operationId is used).

> - For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the api-name along the `operationId` as file name: `device-identifier-retrieveType.feature`.

Comment (jlurien) [reply documentation/API-Testing-Guidelines.md]: I agree. Applied

**Comment by jlurien** [comment]
Final comments addressed. Thanks for the feedback and corrections.

**Review by rartych — APPROVED** [review]
LGTM

**Review by patrice-conil — APPROVED** [review]
LGTM

**Comment by mdomale** [comment]
@jlurien Thanks for accommodating comments .LGTM  now cc @akoshunyadi  @shilpa-padgaonkar

**Review by mdomale — APPROVED** [review]
LGTM thanks

**Review by PedroDiez — APPROVED** [review]
LGTM
